pageserver/src/virtual_file.rs

```diff
@@ -674,14 +674,14 @@ impl VirtualFile {
     /// Returns the IoBuf that is underlying the BoundedBuf `buf`.
     /// I.e., the returned value's `bytes_init()` method returns something different than the `bytes_init()` that was passed in.
     /// It's quite brittle and easy to mis-use, so, we return the size in the Ok() variant.
-    pub async fn write_all<B: BoundedBuf<Buf = Buf>, Buf: IoBuf + Send>(
+    pub async fn write_all<Buf: IoBuf + Send>(
         &mut self,
-        buf: B,
+        buf: Slice<Buf>,
         ctx: &RequestContext,
-    ) -> (B::Buf, Result<usize, Error>) {
+    ) -> (Slice<Buf>, Result<usize, Error>) {
```

Comment:
1. The doc comment is out of date.
2. I think we should return the `Buf` instead of the `Slice`, so, the doc comments aren't out of date if you follow that suggestion.

Reason why we should return the `Buf` instead of the `Slice<Buf>`: the bounds aren't the same when we return, that could mis-lead users of this API.

If you want to keep the bounds, please take inspiration from the doc comment and assertions in `read_exact_at`.

Reply:
We had a 1:1 call and we agreed it is fine to have an API that mirrors the read API as long as the returned slice matches precisely the passed slice (like in the read API, see `read_exact_at`).